rfd/0209-mcp-access.md

```diff
@@ -0,0 +1,510 @@
+---
+authors: STeve Huang (xin.huang@goteleport.com)
+state: draft
+---
+
+# RFD 0209 - MCP Access
+
+## Required Approvers
+
+* Engineering: @r0mant
+* Product: @klizhentas
+
+## What
+
+Support zero-trust access for MCP servers.
+
+## Why
+
+Introduced in late 2024, Anthropic’s [Model Context Protocol
+(MCP)](https://modelcontextprotocol.io/introduction) is a widely adopted,
+open-source standard that enables language models to seamlessly interact with
+external tools and data, enhancing their contextual capabilities.
+
+However, MCP servers today are mostly operated locally without secure transport
+or authorization. OAuth authorization was recently added to the specification,
+but as of the time of writing, it is still new and not widely adopted.
+
+With Teleport's support for MCP servers, users can:
+- Access MCP servers on remote machines where transport is secured with TLS
+- Control access to specific MCP servers and specific tools defined by Teleport roles
+- Track activity through Teleport's audit log
+
+## Scope
+
+Teleport will support stdio-based MCP servers in the initial implementation,
+with [streamable
+HTTP](https://modelcontextprotocol.io/specification/2025-03-26/basic/transports#streamable-http))
+and
+[OAuth](https://modelcontextprotocol.io/specification/2025-03-26/basic/authorization)
+support planned for future iterations (see the "Future development" section for
+details).
+
+## Details
+
+To expedite the initial rollout, the first iteration will focus on adding MCP
+protocol support for Application Access, similar to existing HTTP and TCP
+applications.
+
+### UX
+
+#### editor - configure a filesystem MCP server and define RBAC
+To configure a [filesystem MCP
+server](https://github.com/modelcontextprotocol/servers/tree/main/src/filesystem)
+with `npx` and a [Slack API MCP
+server](https://github.com/modelcontextprotocol/servers/tree/main/src/slack)
+with `docker` via Teleport App service:
+```yaml
+app_service:
+  enabled: true
+  apps:
+  - name: "dev-files"
+    description: "Shared files for developers"
+    labels:
+      env: "dev"
+    mcp:
+      # Command and arguments to launch a stdio-based MCP server per client connection
+      # on the Teleport service host.
+      command: "npx"
+      args: ["-y", "@modelcontextprotocol/server-filesystem", "/home/dev/files"]
+      # Specifies the local user account under which the command will be executed.
+      # Required for stdio-based MCP servers.
+      run_as_local_user: "dev"
```

Comment:
I think you may want to run these commands under different users depending on Teleport user. For example, to ensure that an MCP server only has access to a part of a filesystem available to that user. 

Should we make this a part of the role instead? Similar idea to [host users](https://goteleport.com/docs/enroll-resources/server-access/guides/host-user-creation/#define-a-teleport-role). Then it can also be templated via traits. E.g.

```yaml
spec:
  allow:
    mcp:
      run_as: "{{external.username}}"
      tools:
      - ...
```

Reply:
we could 

though, i suspect what local account to run as is more tied to the kind of MCP server getting launched than the identity launching it. for example, many MCP servers just API wrappers like google maps, brave search.

what @Tener suggested may make more sense to me where we support variable interpolation in the app definition like in `run_as_local_user` and in `args`.

Or maybe we can have the `role.allow.mcp.run_as` to overwrite the one in app definition if specified?

Reply:
The design surface is pretty large here; another option would be to reuse numeric user ID assignments and/or couple this with automatic user provisioning on Linux. I think we should keep it simple for now.

Reply:
updated the runtime section 👍